Expose supplying a type when creating the VM

rbradford opened the pull request:
The KVM_CREATE_VM ioctl can take a parameter that is a platform and
architecture specific VM type. This can be used for SEV or TDX VMs on
x86_64 platforms of for conveying the IPA size on aarch64 platforms.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford committed:
Expose supplying a type when creating the VM

The KVM_CREATE_VM ioctl can take a parameter that is a platform and
architecture specific VM type. This can be used for SEV or TDX VMs on
x86_64 platforms of for conveying the IPA size on aarch64 platforms.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

rbradford commented:
@andreeaflorescu It's failing because it's pulling in two copies of vmm-sys-util. Is that because we need to pin to a newer kvm-bindings?

rbradford commented:
This is with local Cargo.lock deleted:

```
rob@artemis:~/src/kvm-ioctls (2021-02-23-kvm-create-type)$ cargo tree
kvm-ioctls v0.7.0 (/home/rob/src/kvm-ioctls)
├── kvm-bindings v0.3.1
│   └── vmm-sys-util v0.7.0
│       ├── bitflags v1.2.1
│       └── libc v0.2.87
├── libc v0.2.87
└── vmm-sys-util v0.8.0
    ├── bitflags v1.2.1
    └── libc v0.2.87
[dev-dependencies]
└── byteorder v1.4.2
```

lauralt commented:
kvm-bindings v0.3.1 is using vmm-sys-util v0.7.0, while kvm-ioctls  -> vmm-sys-util v.0.8.0 (details [here](https://github.com/rust-vmm/kvm-bindings/pull/31) and in the issue linked there). Looks like we should pin the kvm-bindings to =0.3.0.

rbradford commented:
> kvm-bindings v0.3.1 is using vmm-sys-util v0.7.0, while kvm-ioctls -> vmm-sys-util v.0.8.0 (details [here](https://github.com/rust-vmm/kvm-bindings/pull/31) and in the issue linked there). Looks like we should pin the kvm-bindings to =0.3.0.

@lauralt, I tested that, and it does work but only because the 3.0 version of kvm-bindings is more flexible in the version numbering. However that feels wrong that newer kvm-ioctls (0.7.x) forces a dependency on an older version of kvm-bindings whereas the older version of kvm-ioctls (the 0.6.x) version forces a dependency on a newer version of kvm-bindings. :exploding_head:

rbradford committed:
Depend on kvm-bindings =0.3.0

The newer 0.3.1 release has a fixed dependency on a conflicting version
of vmm-sys-utils so it is necessary to fix this crate to depend on the
older version which has a more flexible vmm-sys-utils dependency.

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

lauralt commented:
> > kvm-bindings v0.3.1 is using vmm-sys-util v0.7.0, while kvm-ioctls -> vmm-sys-util v.0.8.0 (details [here](https://github.com/rust-vmm/kvm-bindings/pull/31) and in the issue linked there). Looks like we should pin the kvm-bindings to =0.3.0.
> 
> @lauralt, I tested that, and it does work but only because the 3.0 version of kvm-bindings is more flexible in the version numbering. However that feels wrong that newer kvm-ioctls (0.7.x) forces a dependency on an older version of kvm-bindings whereas the older version of kvm-ioctls (the 0.6.x) version forces a dependency on a newer version of kvm-bindings. exploding_head

Yeah, it is indeed very weird. I am not sure what would be the best path forward here. Maybe release a new kvm-bindings version (v0.4.0). :(
